### PR TITLE
Alternate indexing

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/repository/PrisonerBRepository.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.justice.digital.hmpps.prisonersearch.repository
+
+import org.springframework.data.domain.Page
+import org.springframework.data.domain.Pageable
+import org.springframework.data.elasticsearch.annotations.Query
+import org.springframework.data.elasticsearch.repository.ElasticsearchRepository
+import uk.gov.justice.digital.hmpps.prisonersearch.model.Prisoner
+import uk.gov.justice.digital.hmpps.prisonersearch.model.PrisonerB
+import java.time.LocalDate
+
```
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/repository/PrisonerBRepository.kt:10]: For now having to double up the repo and class to provide the out of box spring data support - May be a better way!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/Prisoner.kt [hunk 2]
```diff
@@ -1,40 +1,14 @@
...
-        offenderBooking.bookingNo,
-        offenderBooking.firstName,
-        offenderBooking.lastName,
-        offenderBooking.dateOfBirth,
-        offenderBooking.agencyId,
-        offenderBooking.active
-    )
\ No newline at end of file
+interface Prisoner {
+  var prisonerId: String?
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/Prisoner.kt:6]: is prisoner id, offender number then?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/config/LocalStackConfig.kt [hunk 3]
```diff
@@ -20,7 +25,18 @@ class LocalStackConfig {
   }
 
   @Bean
-  fun localStackContainer(): LocalStackContainer {
+  fun elasticSearchContainer() : ElasticsearchContainer {
+    return ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch-oss:6.7.2")
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/config/LocalStackConfig.kt:29]: So why do we need localstack test containers and this special elastic search container?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/repository/PrisonerBRepository.kt (hunk 1 above)
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/repository/PrisonerBRepository.kt:10]: Yes @andymarke I'll have a go at that in next commit

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/Prisoner.kt (hunk 2 above)
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/Prisoner.kt:6]: I'm trying to use the terminology we use in the front end of our applications - but I think we need to agree on this.  Maybe offender no is better

File: docker-compose-circleci.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  localstack:
+    image: localstack/localstack:0.10.5
+    networks:
+      - hmpps
+    container_name: localstack
+    ports:
+      - "4566-4597:4566-4597"
+      - 8999:8080
+    environment:
+      - SERVICES=sqs,sns,es
```
[the remark this answers is not in the document]
Comment (Mjwillis) [reply docker-compose-circleci.yml:12]: yes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/PrisonerB.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.data.annotation.Id
+import org.springframework.data.elasticsearch.annotations.DateFormat
+import org.springframework.data.elasticsearch.annotations.Document
+import org.springframework.data.elasticsearch.annotations.Field
+import org.springframework.data.elasticsearch.annotations.FieldType
+import uk.gov.justice.digital.hmpps.prisonersearch.services.OffenderBooking
+import java.time.LocalDate
+
+@Document(indexName = "prisoner-search-b")
+class PrisonerB(
```
[the remark this answers is not in the document]
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/model/PrisonerB.kt:12]: Didn't seem to work, but going to relook

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResource.kt [hunk 6]
```diff
@@ -16,33 +17,35 @@ import org.springframework.web.bind.annotation.RestController
...
 
-    @GetMapping("/findBy/prisonerId/{prisonerId}")
-    @ApiOperation(value = "Find by Prisoner Id", authorizations = [Authorization("GLOBAL_SEARCH")])
-    @PreAuthorize("hasAnyRole('GLOBAL_SEARCH','SYSTEM_USER')")
-    fun findByPrisonerId(@ApiParam("Prisoner Id", example = "A1234AA") @PathVariable @Size(max = 7) prisonerId: String): Prisoner {
-        return prisonerSearchService.findByPrisonerId(prisonerId)
+    @GetMapping("/find-by/id/{id}")
+    @ApiOperation(value = "Find by IDs")
+    @PreAuthorize("hasAnyRole('GLOBAL_SEARCH')")
+    fun findByPrisonerId(@ApiParam("id", example = "A1234AA", required = true) @PathVariable id: String): Prisoner? {
```
[the remark this answers is not in the document]
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerSearchResource.kt:29]: yes

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/config/LocalStackConfig.kt (hunk 3 above)
Comment (Mjwillis) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/config/LocalStackConfig.kt:29]: test container for localstack doesn't have ES

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerIndexResource.kt [hunk 7]
```diff
@@ -1,37 +1,32 @@
...
+        return prisonerIndexService.buildIndex()
     }
 
-    @GetMapping("/rebuild")
-    @ApiOperation(value = "Index everything", authorizations = [Authorization("PRISONER_INDEX")])
-    @PreAuthorize("hasAnyRole('PRISONER_INDEX','SYSTEM_USER')")
-    fun rebuild(): Map<String, String> {
-        val numberIndexed = prisonerIndexService.indexAll()
-        return mapOf("number-indexed" to numberIndexed.toString())
+    @PutMapping("/mark-complete")
```
[the remark this answers is not in the document]
Comment (Mjwillis) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/PrisonerIndexResource.kt:25]: Yes when index is finished it swaps over and stops double updating.

**Pull request merged**